gateway/ln-gateway/src/lib.rs

```diff
@@ -950,18 +956,34 @@ impl Gateway {
         Err(GatewayError::Disconnected)
     }
 
-    pub async fn handle_leave_federation(&mut self, payload: LeaveFedPayload) -> Result<()> {
+    pub async fn handle_leave_federation(
+        &mut self,
+        payload: LeaveFedPayload,
+    ) -> Result<FederationInfo> {
         // TODO: This should optimistically try to contact the federation to remove the
         // registration record
-        self.remove_client(payload.federation_id).await?;
+        let client = self.remove_client(payload.federation_id).await?;
+
+        let federation_info = self
+            .make_federation_info(client.value(), payload.federation_id)
+            .await;
+
+        // Remove previously assigned scid from `scid_to_federation` map
+        self.scid_to_federation
+            .write()
+            .await
+            .retain(|_, fid| *fid != payload.federation_id);
```

Comment:
I'm confused by this. Why are we retaining it? I think we should ideally remove it from `scid_to_federation` when we leave the federation, I'm surprised we didn't do this already in `remove_client`.

Is this so we can track short channel ids from federations that we've left?

Reply:
the predicate in retain has `!=` to make sure we remove this scid. I know. Looks confusing at fist glance

Reply:
I don't think I like the idea of using `scid_to_federation` has the way of keeping around old short channel ids. If I'm reading this correctly, when the gateway reboots, it re-reads the short channel ids from the configs that are persisted in the database. In this case, since previously the gateway just left a federation, the config will no longer be there, and the `scid_to_federation` will not contain this federation either. Is this the intended behavior?

Reply:
I think the function name `retain` is confusing, but the predicate logic actually removes the scid that was previously assigned to the federation client we are removing.
> I think we should ideally remove it from scid_to_federation when we leave the federation

It results in this behavior you described, i.e. `scid_to_federation` only retains scids that are assigned to currently connected federations, even beyond gateway reboot

Reply:
Got it. Can we move this into `remove_client`? Seems to be the more natural spot to me

Reply:
Agreed that is more natural. It was challenging because we need to `make_federation_info` with the client we just removed. I did it here as a convenience so I can still use the make federation helper, then it seemed like a reasonable fit because we also do other irreversible cleanup operations here like db updates

Reply:
Can we call `make_federation_info` first then remove the client? I think this should be in `remove_client`, we call `remove_client` in our tests in addition to here and it seems wrong to me that that function doesn't remove it from `scid_to_federation` map

Reply:
Naa, we should revive #4266  and delete `remove_client`. I'll make an attempt of it, picking off from @justinmoon 's investigation

Reply:
@m1sterc001guy, should we go in this direction?  https://github.com/fedimint/fedimint/pull/4266#issuecomment-1936778256

Reply:
I'm open to a workaround for our tests so that we can use `select_client` instead of `remove_client`, but I think that will take some time. I think we should move this logic into `remove_client` for now and if we need we can move it back out when we have a fix for our tests that allows us to delete `remove_client`

Reply:
refactored :)